src/clib/solve_rate_cool_g-cpp.C

```diff
@@ -0,0 +1,1012 @@
+///
+/// @param nelem The number of elements a buffer is expected to have in order
+///    to store values for the standard sized index-range
+SpeciesRateSolverScratchBuf new_SpeciesRateSolverScratchBuf(int nelem) {
+  SpeciesRateSolverScratchBuf out;
+
+  out.ddom = (double*)malloc(sizeof(double)*nelem);
+
+  out.dedot = (double*)malloc(sizeof(double)*nelem);
+  out.HIdot = (double*)malloc(sizeof(double)*nelem);
+  out.dedot_prev = (double*)malloc(sizeof(double)*nelem);
+  out.HIdot_prev = (double*)malloc(sizeof(double)*nelem);
+
+  out.k13dd = (double*)malloc(sizeof(double)*nelem*14);
+
+  out.h2dust = (double*)malloc(sizeof(double)*nelem);
+
+  out.itmask_nr = (gr_mask_type*)malloc(sizeof(gr_mask_type)*nelem);
+
+  out.imp_eng = (int*)malloc(sizeof(int)*nelem);
+
+  out.species_tmpdens = grackle::impl::new_SpeciesCollection(nelem);
+  out.kcr_buf = grackle::impl::new_CollisionalRxnRateCollection(nelem);
+  out.kshield_buf = grackle::impl::new_PhotoRxnRateCollection(nelem);
+  out.chemheatrates_buf = grackle::impl::new_ChemHeatingRates(nelem);
+  out.grain_growth_rates = grackle::impl::new_GrainSpeciesCollection(nelem);
+
+  return out;
+}
+
+/// performs cleanup of the contents of SpeciesRateSolverScratchBuf
+///
+/// This effectively invokes a destructor
+void drop_SpeciesRateSolverScratchBuf(SpeciesRateSolverScratchBuf* ptr) {
+  free(ptr->ddom);
+
+  free(ptr->dedot);
+  free(ptr->HIdot);
+  free(ptr->dedot_prev);
+  free(ptr->HIdot_prev);
+
+  free(ptr->k13dd);
+
+  free(ptr->h2dust);
+
+  free(ptr->itmask_nr);
+  free(ptr->imp_eng);
+
+  grackle::impl::drop_SpeciesCollection(&ptr->species_tmpdens);
+  grackle::impl::drop_CollisionalRxnRateCollection(&ptr->kcr_buf);
+  grackle::impl::drop_PhotoRxnRateCollection(&ptr->kshield_buf);
+  grackle::impl::drop_ChemHeatingRates(&ptr->chemheatrates_buf);
+  grackle::impl::drop_GrainSpeciesCollection(&ptr->grain_growth_rates);
+}
+
+
+} // namespace grackle::impl
+
+// -------------------------------------------------------------
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+int solve_rate_cool_g(
+  int imetal, double dt, InternalGrUnits internalu,
+  chemistry_data* my_chemistry, chemistry_data_storage* my_rates,
+  grackle_field_data* my_fields, photo_rate_storage* my_uvb_rates
+)
+{
+  // shorten `grackle::impl::fortran_wrapper` to `f_wrap` within this function
+  namespace f_wrap = ::grackle::impl::fortran_wrapper;
+
+#ifdef GRACKLE_FLOAT_4
+  const gr_float tolerance = (gr_float)(1.0e-05);
+#else
+  const gr_float tolerance = (gr_float)(1.0e-10);
+#endif
+
+  // Set error indicator (we will return this value)
+  int ierr = GR_SUCCESS;
+
+  // Set flag for dust-related options
+  const gr_mask_type anydust =
+    ((my_chemistry->h2_on_dust > 0)  ||  (my_chemistry->dust_chemistry > 0))
+    ? MASK_TRUE
+    : MASK_FALSE;
+
+  // ignore metal chemistry/cooling below this metallicity
+  const double min_metallicity = 1.e-9 / my_chemistry->SolarMetalFractionByMass;
```

Comment:
Let's not change this now since it may require a gold standard update, but I am now fairly certain it should be like this `metallicity` is already in units of Zsun.
```suggestion
  const double min_metallicity = 1.e-9
```